src/bookshelf/src/bksw_wm_sql_object.erl

```diff
@@ -0,0 +1,336 @@
+                 X
+         end,
+    {MT, _Params} = webmachine_util:media_type_to_detail(CT),
+    {[{MT, upload}], Rq, Ctx}.
+
+resource_exists(Rq0, Ctx) ->
+    case wrq:method(Rq0) of
+        %% Buckets always exist for writes since we create them on the fly
+        'PUT' ->
+            case fetch_entry_md(Rq0, Ctx) of
+                {error, _Error} ->
+                    {false, Rq0, Ctx};
+                {not_found, #context{} = Ctx1 } ->
+                    {true, Rq0, Ctx1};
+                {#db_file{}, Ctx1} ->
+                    {true, Rq0, Ctx1}
+            end;
+        _ ->
+            %% determine if the entry exists by opening it. This way, we can cache the fd
+            %% and avoid extra system calls. It also helps to keep the request processing
+            %% more consistent since we will open the fd once at start and hold on to it.
+            %% Note that there is still a possible discrepency when we read the meta data.
+            case fetch_entry_md(Rq0, Ctx) of
+                {error, _Error} ->
+                    {false, Rq0, Ctx};
+                {not_found, #context{} = Ctx1 } ->
+                    {false, Rq0, Ctx1};
+                {#db_file{}, Ctx1} ->
+                    {true, Rq0, Ctx1}
+            end
+    end.
+
+last_modified(Rq0, Ctx) ->
+    case fetch_entry_md(Rq0, Ctx) of
+        {#db_file{created_at = {datetime, Date}}, CtxNew} ->
+            {Date, Rq0, CtxNew};
+        _ ->
+            %% common errors (not found, etc, will have been caught by resource_exists
+            %% because that executes first (G7 in webmachine_decision_core.erl)
+            {halt, Rq0, Ctx}
+    end.
+
+generate_etag(Rq0, Ctx) ->
+    case fetch_entry_md(Rq0, Ctx) of
+        {#db_file{hash_md5 = Digest}, CtxNew} ->
+            {bksw_format:to_base64(Digest), Rq0, CtxNew};
+        _ ->
+            %% common errors (not found, etc, will have been caught by resource_exists
+            %% because that executes first (G7 in webmachine_decision_core.erl)
+            {halt, Rq0, Ctx}
+    end.
+
+delete_resource(Rq0, Ctx) ->
+    case fetch_entry_md(Rq0, Ctx) of
+        {#db_file{file_id = FileId}, CtxNew} ->
+            case bksw_sql:delete_file(FileId) of
+                ok ->
+                    {true, Rq0, CtxNew};
+                _ ->
+                    {halt, Rq0, CtxNew}
+            end;
+        _ ->
+            %% common errors (not found, etc, will have been caught by resource_exists
+            %% because that executes first (G7 in webmachine_decision_core.erl)
+            {halt, Rq0, Ctx}
+    end.
+
+maybe_retry(#context{sql_retry_count = 0}, _RetryFun, FailFun) ->
+    FailFun();
+maybe_retry(#context{sql_retry_delay = Delay, sql_retry_count = Count} = Ctx, RetryFun, _FailFun) ->
+    timer:sleep(Delay),
+    RetryFun(Ctx#context{sql_retry_count = Count - 1}).
+
+%% Return `{Obj, CtxNew}' where `Obj' is the entry meta data `#object{}' record or the atom
```

Comment:
Is this comment out of date? We're returning a `#db_file{}` record not an `#object{}` record. Should we add a dialyzer spec for this?